Address comments for PR 6

JakeLin opened the pull request:
In this PR, I addressed the comments for PR #6 

@adelRestom I have addressed all your comments except for the App.js.

>App.js
>
> 1. Very important thing we usually do is to protect the user from losing Ether on function calls that would fail; we do that by first calling the function with .call() which runs it locally/simulation (locally => no transaction => no mining => no gas consumed => no Ether lost), if that call doesn't fail then we proceed to actually call the function (i.e. call it without .call.
> 2. You used on('transactionHash'), also use on('receipt') to inspect the receipt status (that it's == 1) and to inform the user of a successful transaction

For the first comment, do you call all the `send` method like
`await contract.methods.split(beneficiary1Address, beneficiary2Address).call();`
before
```
await contract.methods.split(beneficiary1Address, beneficiary2Address).send({
  from: accounts[0],
  value: web3.utils.toWei(etherToSplit, 'ether')
});
```

For the second comment, I have checked the `receipt` in the promise call - see `tx` variable, please.

```
  const tx = await contract.methods.split(beneficiary1Address, beneficiary2Address).send({
        from: accounts[0],
        value: web3.utils.toWei(etherToSplit, 'ether')
      }).on('transactionHash', txHash => {
        this.setState({ splitMessage: `Transaction ${txHash} submitted, waiting for the network to mine` });
      }).on('receipt', receipt => {
        console.log(receipt);
      });
      console.log(tx);
```

I print out `receipt` and `tx`, they are exactly same.

@xavierlepretre can you please have a look as well? thanks

JakeLin committed:
[Contract] - Add `changeOwner` to `Ownable` Contract.

JakeLin committed:
[Contract] - Add the parameter `paused` to `Pausable` contract

JakeLin committed:
[Contract] - rename `whenNotPaused` to `whenRunning`

JakeLin committed:
[Contract] - rename `unpause`to `resume` in `Pausable` contract

JakeLin committed:
[Contract] - Use modifier `whenRunning` in  the `pause` method

JakeLin committed:
[Contract] - add `Killable` contract

JakeLin committed:
[Contract] - add `withdrawAll` method in `Splitter` contract

JakeLin committed:
[Contract] - Credit the remainder to the sender's balance when split odd number of ether.

xavierlepretre commented:
What Adel meant is:

```javascript
const result = await contract.methods.split(beneficiary1Address, beneficiary2Address).call(...
if (result) {
    do the .send
```

xavierlepretre commented:
As for the receipt, I see you are checking it status 👍 . You could also check the presence of the expected event.

JakeLin commented:
@xavierlepretre @adelRestom 

I have tried out the following code
```
   const result = await contract.methods.split(beneficiary1Address, beneficiary2Address).call({
        from: accounts[0],
        value: web3.utils.toWei(etherToSplit, 'ether')
      });
      console.log(result);
      const tx = await contract.methods.split(beneficiary1Address, beneficiary2Address).send({
        from: accounts[0],
        value: web3.utils.toWei(etherToSplit, 'ether')
      }).on('transactionHash', txHash => {
        this.setState({ splitMessage: `Transaction ${txHash} submitted, waiting for the network to mine` });
      });
      if (tx.status) {
        const contractBalance = web3.utils.fromWei(await web3.eth.getBalance(contractAddress));
        this.setState({ splitMessage: `Transaction ${tx.transactionHash} got mined successfully.`, contractBalance });
      } else {
        this.setState({ splitMessage: `Something went wrong with transaction ${tx.transactionHash}.` });
      }
```

I found out whether the transaction fails or not in `await contract.methods.split(beneficiary1Address, beneficiary2Address).send(...` , `await contract.methods.split(beneficiary1Address, beneficiary2Address).call(...` will always returns an object of `Result` like
<img width="402" alt="Screen Shot 2019-08-18 at 11 32 14 am" src="https://user-images.githubusercontent.com/573856/63219010-dc3c0480-c1ab-11e9-9953-5ea571909e65.png">

Can you guys tell me how can I meet  @adelRestom 's requirement?

adelrustum commented on contracts/Splitter.sol:
```diff
@@ -33,4 +33,12 @@ contract Splitter is Pausable, Killable {
+
+  function withdrawAll() external onlyOwner whenKilled {
+    uint256 balanceToWithdraw = address(this).balance;
+    require(balanceToWithdraw > 0, "Balance must be grater than zero to withdraw!");
```
I think this line of code becomes irrelevant because the normal withdraw is locked now.

adelrustum commented on contracts/Killable.sol:
```diff
@@ -0,0 +1,23 @@
+  function isKilled() public view returns (bool) {
+    return killed;
+  }
+
```
Make this functionality part of the Pausable contract, we want to protect the user from accidentally calling this function by turning it into a 2 step-process:
1. The user pauses the contract
2. The user kills the contract (the kill function checks that the contract is paused)

adelrustum commented on contracts/Pausable.sol:
```diff
@@ -27,8 +27,8 @@ contract Pausable is Ownable {
 
-  function unpause() public onlyOwner {
-    require(paused, "Can't unpause a non-paused contract!");
+  function resume() public onlyOwner {
```
Add a modifier called `whenPaused`; you'll need it for "kill" function anyway.

adelrustum commented on contracts/Splitter.sol:
```diff
@@ -11,6 +11,9 @@ contract Splitter is Pausable {
 
   mapping (address => uint256) public balances;
```
You'll also need to update your migration script and test file to pass the value to the constructor when calling `new()`.

adelRestom commented:
Regarding the `.call()` returned object, I thought it would throw an exception for a failed call that you can catch with a `.catch()`. 
Try to change your `split` function so that it returns a boolean and add `return true;` at the end of the function, this way you can check the returned boolean value so it would be like Xavier's code (`let result = contract.func.call; if result == true..then call send`).

JakeLin commented:
@adelRestom I have followed your advice to add `return true` in the function like

```
  function split(address _beneficiary1, address _beneficiary2) external payable whenRunning whenAlive returns(bool) {
    require(_beneficiary1 != address(0) && _beneficiary2 != address(0), "Beneficiary's address must not be zero!");
    require(msg.value > 0, "Must split more than zero ether!");

    if (msg.value.mod(2) == 1) {
      balances[msg.sender] = balances[msg.sender].add(1);
    }
    uint256 half = msg.value.div(2);
    balances[_beneficiary1] = balances[_beneficiary1].add(half);
    balances[_beneficiary2] = balances[_beneficiary2].add(half);
    emit LogSplitted(msg.sender, msg.value, _beneficiary1, _beneficiary2);
    return true;
  }
```

And add the `call()` check in App.js like

```
      const result = await contract.methods.split(beneficiary1Address, beneficiary2Address).call({
        from: accounts[0],
        value: web3.utils.toWei(etherToSplit, 'ether')
      });
      if (!result) {
        return;
      }
```

However, even I send 0 wei to the contract, `result` is still `true`. I also tried the same approach in `withdraw` function. They always return `true` when I call `call()`.

So I don't know how can I check before I `send()`.

adelRestom commented:
@xavierlepretre can you please remind me how we used to do this? ☝️

xavierlepretre commented:
I cloned your repo, checked out this branch, replaced 9545, npm installed both folders, truffle migrated into Ganache, npm started.
When I open in the browser, I get:

```
index.js:1375 TypeError: contract.methods.owner is not a function
    at App.componentDidMount (App.js:64)
console.<computed> @ index.js:1375
```

xavierlepretre commented:
I would try to `console.log(typeof result)` and `console.log(JSON.stringify(result))` to collect more info.

JakeLin commented:
@xavierlepretre I just merged the PR, can you rebase master?

adelRestom commented:
I'm looking at my old work and I had something like:
```javascript
return myContract.myFunction.call()
.then(() => myContract.myFunction())
.then...more code
.catch(error => console.log(error));
```
So the catch would show the error.

JakeLin commented:
thanks @adelRestom I am working on project 2 and 3 to try to make meet the deadline, will come back to this soon.

xavierlepretre commented:
I still have the same error when loading the web page:

```
index.js:1375 TypeError: contract.methods.owner is not a function
    at App.componentDidMount (App.js:64)
console.<computed> @ index.js:1375
```

JakeLin commented:
@xavierlepretre sorry, for some reason, I didn't commit the fix to github. I have pushed the commit 522d3d8e36ad3d0a44f955270a34f20351de7ebd now, it should be OK now.

xavierlepretre commented:
Now I get a revert on withdraw.

JakeLin commented:
That's weird, may I come back to this later? I am trying to finish the other projects
